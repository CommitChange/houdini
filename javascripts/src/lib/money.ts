// License: LGPL-3.0-or-later
// based upon https://github.com/davidkalosi/js-money
import BigNumber from 'bignumber.js';

const assertSameCurrency = function (left: any, right: any) {
  if (left.currency !== right.currency)
    throw new Error('Different currencies');
};

const assertType = function (other: any) {
  if (!(other instanceof Money))
    throw new TypeError('Instance of Money required');
};

const assertOperand = function (operand: number|BigNumber) {
  if (
    (typeof operand === 'number' && (isNaN(operand)|| !isFinite(operand))) ||
    (operand instanceof BigNumber && (operand.isNaN() || !operand.isFinite())))

    throw new TypeError('Operand must be a valid number');
};

/**
 * Represents a monetary amount. For safety, all Money objects are immutable. All of the functions in this class create a new Money object.
 * 
 * To create a new Money object is to use the `fromCents` function.
 * @export
 * @class Money
 */
export class Money {

  readonly currency:string
  readonly amountInCents:BigNumber
  protected constructor(amountInCents: number|BigNumber, currency: string) {

    if (typeof amountInCents === 'number'){
      amountInCents = new BigNumber(amountInCents);
    }
    if (!amountInCents.isInteger()) {
      throw new TypeError("Money may only be in whole units");
    }
    this.amountInCents = amountInCents;
    this.currency = currency.toLowerCase();
    Object.bind(this.equals)
    Object.bind(this.add)
    Object.bind(this.subtract)
    Object.bind(this.multiply)
    Object.bind(this.divide)
    Object.bind(this.compare)
    Object.bind(this.greaterThan)
    Object.bind(this.greaterThanOrEqual)
    Object.bind(this.lessThan)
    Object.bind(this.lessThanOrEqual)
    Object.bind(this.isZero)
    Object.bind(this.isPositive)
    Object.bind(this.isNegative)
    Object.bind(this.toJSON)
    Object.bind(this.toString)

    Object.freeze(this);
  }

  /**
   * Create a `Money` object with the given number of cents and the ISO currency unit
   * @static
   * @param  {number} amountInCents 
   * @param  {string} currency 
   * @return Money 
   * @memberof Money
   */
  static fromCents(amountInCents: number|BigNumber, currency: string): Money {
    return new Money(amountInCents, currency)
  }

  /**
   * Create a `Money` object with the given number if smallest monetary units and the ISO currency. Another name for the `fromCents` function.
   * @static
   * @memberof Money
   */
  static fromSMU=Money.fromCents

  /**
  * Returns true if the two instances of Money are equal, false otherwise.
  *
  * @param {Money} other
  * @returns {Boolean}
  */
  equals(other: Money): boolean {
    assertType(other);

    return this.amountInCents.isEqualTo(other.amountInCents) &&
      this.currency === other.currency;
  };

  /**
   * Adds the two objects together creating a new Money instance that holds the result of the operation.
   *
   * @param {Money} other
   * @returns {Money}
   */
  add(other: Money|number): Money {
    if (typeof other === 'number') {
      other = Money.fromCents(other, this.currency);
    }
    assertType(other);
    assertSameCurrency(this, other);

    return new Money(this.amountInCents.plus(other.amountInCents), this.currency);
  };

  /**
   * Subtracts the two objects creating a new Money instance that holds the result of the operation.
   *
   * @param {Money} other
   * @returns {Money}
   */
  subtract(other: Money): Money {
    assertType(other);
    assertSameCurrency(this, other);

    return new Money(this.amountInCents.minus(other.amountInCents), this.currency);
  };

  /**
   * Multiplies the object by the multiplier returning a new Money instance that holds the result of the operation.
   *
   * @param {Number} multiplier
   * @returns {Money}
   */
  multiply(multiplier: number|BigNumber, roundingMode:BigNumber.RoundingMode=BigNumber.ROUND_CEIL): Money {

    assertOperand(multiplier);
    const amount = new BigNumber(this.amountInCents.multipliedBy(multiplier).toFixed(0, roundingMode));

    return new Money(amount, this.currency);
  };

  /**
   * Divides the object by the multiplier returning a new Money instance that holds the result of the operation.
   *
   * @param {Number} divisor
   * @returns {Money}
   */
  divide(divisor: number|BigNumber|Money, roundingMode: BigNumber.RoundingMode = BigNumber.ROUND_CEIL): Money {

    if (divisor instanceof Money) {
      assertSameCurrency(this, divisor)
      divisor = divisor.amountInCents
    }
    
    assertOperand(divisor);
    const oldRoundingMode = BigNumber.config({})
    BigNumber.config({ROUNDING_MODE: roundingMode, DECIMAL_PLACES: 0})
    const amount = this.amountInCents.div(divisor);
    BigNumber.config({...oldRoundingMode})

    return new Money(amount, this.currency);
  };

  /**
   * Allocates fund bases on the ratios provided returing an array of objects as a product of the allocation.
   *
   * @param {Array} other
   * @param {Money[]}
   */
  // allocate(ratios: number[]): Money[] {
  //   var self = this;
  //   var remainder = self.amountInCents;
  //   var results: Money[] = [];
  //   var total = 0;

  //   ratios.forEach(function (ratio) {
  //     total += ratio;
  //   });

  //   ratios.forEach(function (ratio) {
  //     var share = Math.floor(self.amountInCents * ratio / total)
  //     results.push(new Money(share, self.currency));
  //     remainder -= share;
  //   });

  //   for (var i = 0; remainder > 0; i++) {
  //     results[i] = new Money(results[i].amountInCents + 1, results[i].currency);
  //     remainder--;
  //   }

  //   return results;
  // };

  /**
   * Compares two instances of Money.
   *
   * @param {Money} other
   * @returns {Number}
   */
  compare(other: Money): number {

    assertType(other);
    assertSameCurrency(this, other);

    return this.amountInCents.comparedTo(other.amountInCents)
  };

  /**
   * Checks whether the value represented by this object is greater than the other.
   *
   * @param {Money} other
   * @returns {boolean}
   */
  greaterThan(other: Money): boolean {
    assertType(other);
    assertSameCurrency(this, other);
    return this.amountInCents.isGreaterThan(other.amountInCents)
  };

  /**
   * Checks whether the value represented by this object is greater or equal to the other.
   *
   * @param {Money} other
   * @returns {boolean}
   */
  greaterThanOrEqual(other: Money): boolean {
    assertType(other);
    assertSameCurrency(this, other);
    return this.amountInCents.isGreaterThanOrEqualTo(other.amountInCents)
  };

  /**
   * Checks whether the value represented by this object is less than the other.
   *
   * @param {Money} other
   * @returns {boolean}
   */
  lessThan(other: Money): boolean {
    assertType(other);
    assertSameCurrency(this, other);
    return this.amountInCents.isLessThan(other.amountInCents);
  };

  /**
   * Checks whether the value represented by this object is less than or equal to the other.
   *
   * @param {Money} other
   * @returns {boolean}
   */
  lessThanOrEqual(other: Money): boolean {
    return this.amountInCents.isLessThanOrEqualTo(other.amountInCents)
  };

  /**
   * Returns true if the amount is zero.
   *
   * @returns {boolean}
   */
  isZero(): boolean {
    return this.amountInCents.isZero();
  };

  /**
   * Returns true if the amount is positive.
   *
   * @returns {boolean}
   */
  isPositive(): boolean {
    return this.amountInCents.isPositive();
  };

  /**
   * Returns true if the amount is negative.
   *
   * @returns {boolean}
   */
  isNegative(): boolean {
    return this.amountInCents.isNegative();
  };

  /**
   * Returns a serialised version of the instance.
   *
   * @returns {{amount: number, currency: string}}
   */
  toJSON(): { amountInCents: number; currency: string; } {
    return {
      amountInCents: this.amountInCents.toNumber(),
      currency: this.currency
    };
  };

  toString(): string {
    return `${this.amountInCents.toNumber().toString()} ${this.currency}`;
  }
}

export default Money;