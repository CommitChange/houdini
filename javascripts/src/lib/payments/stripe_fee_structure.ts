// License: LGPL-3.0-or-later
import {boundMethod} from 'autobind-decorator'
import { Money } from "../money";
import _ = require('lodash');

type FeeStructureProps = {flatFee?:number, percentFee?:number}

export class StripeFeeStructure {

    constructor(readonly props:FeeStructureProps)
    {
        if (!props.flatFee && !props.percentFee) {
            throw Error(`flatFee or percentFee must be passed`)
        }

        const def:FeeStructureProps = {
            flatFee: 0,
            percentFee: 0
        }

        props = {...def, ...props}

        if (props.percentFee < 0 || props.percentFee > 1)
        {
            throw Error(`percentFee must be less or equal to 1 and greater then or equal to 0`)
        }

        if (props.flatFee < 0) {
            throw Error(`flatFee must be 0 or greater`)
        }

        if (!_.isInteger(props.flatFee)){
            throw Error(`flatFee must be an integer as its always in the smallest currency unit`)
        }

        this.props = props;
        Object.freeze(this.props)
    }
    
    @boundMethod
    calculateFee(x:Money) : Money {
        return Money.fromCents(Math.round(x.amountInCents - (x.amountInCents - this.props.flatFee - (x.amountInCents * this.props.percentFee))), x.currency);
    }

    @boundMethod
    reverseCalculateFee(x:Money):Money {
        return Money.fromCents(Math.round(((x.amountInCents + this.props.flatFee) / (1 - this.props.percentFee) - x.amountInCents)), x.currency)
    }
}