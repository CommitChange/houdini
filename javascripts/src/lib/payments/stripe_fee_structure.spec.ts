// License: LGPL-3.0-or-later
import { StripeFeeStructure } from './stripe_fee_structure'
import 'jest';
import { Money } from '../money';

describe('StripeFeeStructure', () => {
  describe('constructor', () => {
    it('throws on missing arguments', () => {
      const sfs = StripeFeeStructure;
      expect(() => new sfs({})).toThrowError(Error)
    })

    it('throws when flatFee is less than zero', () => {
      expect(() => new StripeFeeStructure({flatFee: -1})).toThrowError()
    })

    it('throws when percentFee is less than zero', () => {
      expect(() => new StripeFeeStructure({percentFee: -1})).toThrowError()
    })

    it('throws when percentFee is greater than 1', () => {
      expect(() => new StripeFeeStructure({percentFee: 1.1})).toThrowError()
    })

    it('it throws when flatFee is not an integer', () => {
      expect(() => new StripeFeeStructure({flatFee: 2.2})).toThrowError()
    })

    it('returns the expected structure when passing both', () => {
      expect(new StripeFeeStructure({percentFee: .02, flatFee: 30})).toMatchSnapshot()
    })

    it('returns the expected structure when passing percentFee', () => {
      expect(new StripeFeeStructure({percentFee: .02})).toMatchSnapshot()
    })

    it('returns the expected structure when passing flatFee', () => {
      expect(new StripeFeeStructure({flatFee: 30})).toMatchSnapshot()
    })
  })

  describe('calculateFee', () => {
    const feeStructure = new StripeFeeStructure({percentFee: .022, flatFee: 30})
    it('entering 100 gets fee of 32', () => {
      expect(feeStructure.calculateFee(Money.fromCents(100, 'USD'))).toEqual(Money.fromCents(32, 'usd'));
    })

    it('entering 10000 gets fee of 250', () => {
      expect(feeStructure.calculateFee(Money.fromCents(10000, 'USD'))).toEqual(Money.fromCents(250, 'usd'))
    })
  })

  describe('reverseCalculateFee', () => {
    const feeStructure = new StripeFeeStructure({percentFee: .022, flatFee: 30})
    it('entering 100 gets proper fee', () => {

      const oneHundred = Money.fromCents(100, 'USD')
      let calcFee = feeStructure.reverseCalculateFee(oneHundred)

      expect(feeStructure.calculateFee(calcFee.add(oneHundred))).toEqual(calcFee)
    })

    it('entering 10000 gets proper fee', () => {
      const tenThousand = Money.fromCents(10000, 'USD')
      let calcFee = feeStructure.reverseCalculateFee(tenThousand)

      expect(feeStructure.calculateFee(calcFee.add(tenThousand))).toEqual(calcFee)
    })
  })
})