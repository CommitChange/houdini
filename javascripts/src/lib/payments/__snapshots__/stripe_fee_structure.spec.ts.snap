// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StripeFeeStructure constructor returns the expected structure when passing both 1`] = `
StripeFeeStructure {
  "props": Object {
    "flatFee": 30,
    "percentFee": 0.02,
  },
}
`;

exports[`StripeFeeStructure constructor returns the expected structure when passing flatFee 1`] = `
StripeFeeStructure {
  "props": Object {
    "flatFee": 30,
    "percentFee": 0,
  },
}
`;

exports[`StripeFeeStructure constructor returns the expected structure when passing percentFee 1`] = `
StripeFeeStructure {
  "props": Object {
    "flatFee": 0,
    "percentFee": 0.02,
  },
}
`;
